### ISSUE TITLE
TypeError: this.opts.wsEngine is not a constructor when creating socket connection on server

### ISSUE THREAD

**Issue by jzybert** [socketio/socket.io#3859]
**Describe the bug**
When I created a socket connection for my server, I see the following error:
```
App threw an error during load
TypeError: this.opts.wsEngine is not a constructor
    at Server.init (webpack:///./node_modules/engine.io/lib/server.js?:81:15)
    at new Server (webpack:///./node_modules/engine.io/lib/server.js?:68:10)
    at Function.attach (webpack:///./node_modules/engine.io/lib/engine.io.js?:124:18)
    at Server.initEngine (webpack:///./node_modules/socket.io/dist/index.js?:187:27)
    at Server.attach (webpack:///./node_modules/socket.io/dist/index.js?:174:14)
    at new Server (webpack:///./node_modules/socket.io/dist/index.js?:70:18)
    at eval (webpack:///./server/controller.ts?:35:10)

/server/controller.ts is my TypeScript file which creates an express app and uses that to create the socket connection. This error is thrown when I am bundling with webpack.
```

**To Reproduce**

Please fill the following code example:

Socket.IO server version: `4.0.0`

*Server*

```js
import express from 'express';
import {createServer} from 'http';
import {Server} from 'socket.io';

const app = express();
const server = createServer(app);
const io = new Server(server, {});

let userMsgQueue = new Queue();

io.on("connection", (socket) => {
    console.log('server connected');

    socket.on('message', msg => {
        console.log(`Message: ${JSON.stringify(msg)}`);
    });

    socket.on('disconnect', () => {
        console.log('server disconnected');
    });
});

const PORT = 5000;
server.listen(PORT);
```

Socket.IO client version: `4.0.0`

*Client*

```js
import { io } from "socket.io-client";

let socket = io("http://localhost:5000");

socket.on('connection', () => {
    console.log(`Client connected: ${socket.id}`);
}

socket.on('disconnect', () => {
    console.log("Client disconnected");
});
```

**Expected behavior**
I would expect for this error to not be thrown. I am following the server initialization steps here for v4: https://socket.io/docs/v4/server-initialization/index.html

**Platform:**
 - Device: MacBook Pro (16-inc, 2019)
 - OS: macOS Catalina Version 10.15.7

**Additional context**
Other relevant modules:
```
"express": "^4.17.1",
"typescript": "^4.2.3",
"@babel/cli": "^7.1.0",
"@babel/core": "^7.3.4",
"@babel/plugin-proposal-class-properties": "^7.13.0",
"@babel/plugin-proposal-object-rest-spread": "^7.13.0",
"@babel/plugin-transform-runtime": "^7.13.10",
"@babel/preset-env": "^7.0.0",
"@babel/preset-react": "^7.0.0",
"@babel/preset-typescript": "^7.13.0",
"@types/express": "^4.17.11",
"webpack": "^4.46.0",
"webpack-cli": "^4.5.0",
"webpack-dev-server": "^3.1.3",
```

Here's my webpack config file:
```
const path = require("path");
const webpack = require("webpack");
const OptimizeCssAssetsPlugin = require("optimize-css-assets-webpack-plugin");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const {CleanWebpackPlugin} = require("clean-webpack-plugin");

module.exports = {
    target: 'electron-renderer',
    entry: {
        main: './src/main.ts',
        index: './src/index.tsx'
    },
    module: {
        rules: [
            {
                test: /\.(js|jsx|ts|tsx)$/,
                exclude: /(node_modules)/,
                loader: "babel-loader",
                options: {
                    presets: [
                        "@babel/env",
                        "@babel/react",
                        "@babel/typescript"
                    ],
                    plugins: [
                        "@babel/plugin-transform-runtime",
                        "@babel/proposal-class-properties",
                        "@babel/proposal-object-rest-spread"
                    ]
                }
            },
            {
                test: /\.s[ac]ss$/i,
                use: ["style-loader", "css-loader", "sass-loader"]
            },
            {
                test: /\.css$/,
                use: ["style-loader", "css-loader"]
            },
            {
                test: /\.(jpe?g|png|gif|woff|woff2|eot|ttf)(\?v=\d+\.\d+\.\d+)?$/,
                use: ["file-loader"]
            },
            {
                test: /\.svg$/,
                use: ["react-svg-loader"]
            }
        ]
    },
    resolve: {
        extensions: ["*", ".js", ".jsx", ".ts", ".tsx"]
    },
    node: {
        __dirname: false
    },
    externals: {
        'express': 'commonjs express'
    },
    output: {
        path: path.resolve(__dirname, "dist"),
        filename: "[name].bundle.js",
        globalObject: 'this'
    },
    plugins: [
        new webpack.HotModuleReplacementPlugin(),
        new OptimizeCssAssetsPlugin({}),
        new CleanWebpackPlugin(),
        new HtmlWebpackPlugin({})
    ]
};
```

**Comment by jzybert** [comment]
When I downgrade to version 3.1.2, I see a similar error:
```
App threw an error during load
TypeError: wsModule.Server is not a constructor
    at Server.init (webpack:///./node_modules/engine.io/lib/server.js?:82:15)
    at new Server (webpack:///./node_modules/engine.io/lib/server.js?:66:10)
    at Function.attach (webpack:///./node_modules/engine.io/lib/engine.io.js?:124:18)
    at Server.initEngine (webpack:///./node_modules/socket.io/dist/index.js?:186:27)
    at Server.attach (webpack:///./node_modules/socket.io/dist/index.js?:173:14)
    at new Server (webpack:///./node_modules/socket.io/dist/index.js?:69:18)
    at eval (webpack:///./server/controller.ts?:35:10)
```

It seems like this note about upgrading from v3 to v4 was to address it so perhaps something is missing?
https://socket.io/docs/v4/migrating-from-3-x-to-4-0/#wsEngine-option

**Comment by mjprogramation** [comment]
use on connect instead of connection, and you have some errors in your code formatting, i'm trying to refactor the issue, but not getting any errors:

```
import { io } from "socket.io-client";

let socket = io("http://localhost:5000");

socket.on('connection', () => {
    console.log(`Client connected: ${socket.id}`);
    socket.on('disconnect', () => {
       console.log("Client disconnected");
     }
});   
```

**Comment by jzybert** [comment]
> use on connect instead of connection, and you have some errors in your code formatting, i'm trying to refactor the issue, but not getting any errors:
> 
> ```
> import { io } from "socket.io-client";
> 
> let socket = io("http://localhost:5000");
> 
> socket.on('connection', () => {
>     console.log(`Client connected: ${socket.id}`);
>     socket.on('disconnect', () => {
>        console.log("Client disconnected");
>      }
> });   
> ```

The issue is not in the client code, I only included it to show I was using client as well as server. It’s the server code that is throwing the error. Also connection does work, so that’s not the problem.

**Comment by ralexandr** [comment]
Met same issue :(

**Comment by xinaesthete** [comment]
Me too, in the context of starting from this template https://github.com/cawa-93/vite-electron-builder. My stacktrace looks a bit different as I'm not using WebPack.

**Comment by xinaesthete** [comment]
Digging through the bundled js, it doesn't look as though `wsEngine` ever gets assigned anything other than `DEFAULT_WS_ENGINE = browser.Server`, where
```js
var browser = function() {
  throw new Error("ws does not work in the browser. Browser clients must use the native WebSocket object");
};
```

So perhaps we are each missing some configuration that will cause `socket.io` to be more properly bundled for the server.

**Comment by xinaesthete** [comment]
I have found that if I add `socket.io` to `rollupOptions: { external: [] }`, then it appears to run without error (not sure why I've got some other trivial issue getting it to work usefully).

I have another project using `esbuild` directly, where I'm on an older version `0.8.23`, in that context I don't make `socket.io` external, and seem to have `platform: 'node'` in place of `target: 'node14'`.

I indirectly depend on `esbuild 0.9.7` with vite fwiw.

I don't know exactly how the rollup options relate to the esbuild part of the configuration in vite.

**Comment by xinaesthete** [comment]
**tl;dr:** It's working for me as long as `ws` is external.

(that could be indirectly by making `socket.io` external).

I tried installing `ws` and explicitly passing ws.Server
```js
import {Server} from 'socket.io';
import ws from 'ws';

const io = new Server(server, { wsEngine: ws.Server });
```
but it again complains that `wsEngine is not a constructor` (it does that regardless of whether `ws` is external).

**However**
I have found that I can just make 'ws' external rather than 'socket.io', which is preferable in terms of resulting build size etc. So in the case of that vite-electron template, the best result I've come up with is to explicitly install `ws` as a dependency, include it in the list of `externals` and move `socket.io` along with anything that doesn't need to be `external` into `devDependencies`.

Then when I `npm run compile` I get a node_modules folder in the output app which is somewhat smaller than if I make `socket.io` (and thus all of its dependencies) external.

I imagine it'd be fairly similar for WebPack.

In another context the use of `dependencies` vs `devDependencies` would probably be different.

**Comment by arisa325** [comment]
Met same issue.... :'(

**Comment by xinaesthete** [comment]
> Met same issue.... :'(

> tl;dr: It's working for me as long as ws is external.

**Comment by veaba** [comment]
Met same issue :(

**Comment by RonV-AU** [comment]
This example on socket.io is working - https://socket.io/get-started/private-messaging-part-1/
package.json:     "socket.io-client": "^3.1.1"

**Comment by jzybert** [comment]
I upgraded to 4.1.3 and this warning is no longer present.

**Comment by arisa325** [comment]
Met same issue.... :'(

**Comment by ben-rogerson** [comment]
I upgraded from `3.1.1` > `4.5.1` and still received the `[this.opts.wsEngine is not a constructor` error.
The error no longer shows after I remove the ws option which has been manually set so socket.io can use the default value (`require("ws").Server`):

```diff
io = new socketIO({
-    wsEngine: 'ws',
    //...
}).adapter(createAdapter(pubClient, subClient));
```
